torch/_dynamo/variables/optimizer.py

```diff
@@ -57,15 +57,15 @@ def call_method(
         if name == "_init_group":
             try:
                 py_args, py_kwargs = self.get_python_args(*args, **kwargs)
-                self.value._init_group(*py_args, **py_kwargs)
+                ret_val = self.value._init_group(*py_args, **py_kwargs)
                 self.map_sources_and_install_guards(tx)
                 self.update_list_args(tx, args, kwargs, py_args, py_kwargs)
                 # stash a weak_ptr to optimizer to invalidate code
                 # if the optimizer object dies
                 tx.store_global_weakref(self.get_global_name(), self.value)
                 self.create_finalizer(tx)
 
-                return ConstantVariable.create(None)
+                return ConstantVariable.create(ret_val)
```

Comment:
Is this correct? What if the ret val is not constant?

Reply:
@mlazos how do you suggest we fix this?

Reply:
I guess we can't actually allow `_init_group` to return any variables since we do not want to inspect `_init_group`'s execution?

Reply:
I think we need to add guards to the return values at least. We should assume that in the majority case, `has_sparse_grad` and `has_complex` will not vary across runs, and further more even if they did  we would at most recompile once as they are boolean.

Reply:
Actually, I think we can prove that we do not need to install guards on these as long as changes to dtype / layout will trigger recompilation.

Reply:
These are checked here: https://github.com/pytorch/pytorch/blob/12f97bb2e94d6748d3d39d8f8debefde5b6b8dce/torch/csrc/dynamo/guards.cpp#L50
https://github.com/pytorch/pytorch/blob/12f97bb2e94d6748d3d39d8f8debefde5b6b8dce/torch/csrc/dynamo/guards.cpp#L51

Reply:
Note that export graph will not check the dispatch key and hence layout. 

https://github.com/pytorch/pytorch/blob/12f97bb2e94d6748d3d39d8f8debefde5b6b8dce/torch/_dynamo/guards.py#L695

This is potentially bad @voznesenskym. Could we at least check the layout for export?

Reply:
@mlazos this discussion feels like rearranging deck chairs on the Titanic?  As far as I can see, we are unsoundly calling init_group from inside the Dynamo compiler which may have arbitrary user defined side effects. This seems extremely bad! Is there any extra condition which, eg, limits what can possibly be invoked here?

Reply:
We can write a doc comment on `Optimizer` class that tells people not to perform side effects / return non-tensor meta dependent values from `_init_group` :rofl: 

Currently, `_init_group` is effectful though; we install guards / sources on the state that is modified.

Honestly, I don't think things are too bad, I doubt users will abuse this in too bad of a way. Even if we don't have a way to catch it if they do.

Reply:
This in fact is not unsound at all - currently the only side effects done in _init_group are performed once (even in eager) and set up state on the optimizer object (we own this code). Calling it again is in essence a no-op, which is why we can call this from within dynamo. The reason this was done is because _init_group will take ~5 min to trace with dynamo. Setting up the state in a loop is absolutely horrible for dynamo perf. AFAIK, I don't think _init_group had return values in the optimizers we support. We can just graph break in this case, this is purely a compile time optimization.

Reply:
> AFAIK, I don't think _init_group had return values in the optimizers we support.

I think returned `has_sparse_grad` for a while. (derived from the input tensor layout).

I then added a `has_complex` flag derived from the input tensor dtypes.

But as mentioned above, it is not an issue because the current ret vals of `_init_group` are guarded by the tensor dtype and layout.

We can add some comments to `_init_group` so that users will never try to get it to return something not guarded on tensor metas.

Reply:
> This in fact is not unsound at all - currently the only side effects done in _init_group

Yes, and neither is returning something not derived solely from tensor metas.

I guess we just need to ensure that this invariant is never broken.

Reply:
> What I mean by this is that when we wrap tensor variables in the optimizer object, check if they are complex, and if so add a view_as_real op to the graph.

Hmm, that sounds doable, so we could get rid of the `has_sparse_grad`/`has_complex` flags, which are currently only used to shortcut dynamo tracing to my understanding.

Reply:
If you would like, I could try to take this on (but I'll probably do it tmr)

Reply:
I could be wrong but unfortunately I think those checks are still necessary for eager @janeyx99, but overall I am okay with this plan.

Reply:
> I could be wrong but unfortunately I think those checks are still necessary for eager 

Oh right, because `capturable=True` can still be run from eager :cry: so maybe the plan isn't that good.

Reply:
What we could do is to have a "has_sparse_grad" an "has_complex" _group flag_.

Since this would be in the state, I believe that this is guarded by dynamo too.

Reply:
> This in fact is not unsound at all - currently the only side effects done in _init_group are performed once (even in eager) and set up state on the optimizer object (we own this code). Calling it again is in essence a no-op, which is why we can call this from within dynamo.

I don't buy it.

```
        elif isinstance(value, torch.optim.Optimizer):
            return OptimizerVariable(
                value,
                source=self.source,
                guards=self.make_guards(GuardBuilder.TYPE_MATCH),
            )
```

It looks to me like we are willing to wrap any Optimizer in OptimizerVariable, including user defined ones.

The problem you're dealing with is legitimate, but if I'm right, we need to solve it some other way.

Reply:
Repro: https://gist.github.com/ezyang/55b44f4fc042360327fcb9e32a5af93b

Reply:
> This is potentially bad @voznesenskym. Could we at least check the layout for export?

For sure, we have prior art for export guards for stuff in C++ guards. Should be easy to add.

Reply:
> The problem you're dealing with is legitimate, but if I'm right, we need to solve it some other way.

I skimmed OptimizerVariable and I dont understand the plan for how its meant to work - @mlazos where can I read more.

Reply:
OptimizerVariable exists purely to call _init_group and install guards on optimizer state in dynamo. This is to prevent all optimizers from taking 10 minutes to compile.

Here's the PR that introduced it:
https://github.com/pytorch/pytorch/commit/c46af25bb3d4cd95485682ea5574dd47dab5dd90

Reply:
I agree you are solving a legitimate problem but my point is that the way it is currently implemented is hella unsafe.

Maybe a compromise is to introduce something like `allow_in_graph` to the `_init_group` functions so we explicitly annotate that they are OK to directly run at Dynamo time.

cc @jansel

Reply:
@ezyang note that we do not actually need to run init_group at runtime because it was just a way of sequencing the input states correctly for the graph capture

Reply:
Yeah I was replying to give Voz some background, I'm fine with an explicit marker on `_init_group`. This might actually be useful in other contexts too.

Reply:
cc @yanboliang this is another knob on the INLINE/ALLOW/SKIP determination, feels like.

Reply:
So what's the conclusion on this PR? Do we need to fix the unsound `_init_group` ness first? It's hard for me to reason about the correctness of the PR in this intermediate state.

Reply:
https://github.com/pytorch/pytorch/issues/111209

Reply:
I think that based on the contract of `allow_in_graph`, the general thrust of this PR is fine, but do we need to relocate the code defined here to elsewhere?

Also, how can we lower `_init_group`?

Reply:
Allow in graph is not quite right as we don't actually put these in the graph

Reply:
Well we could do a builtin dynamo lowering of `_init_group` for this particular case, which is to consume it and not write to the tape? 

I guess other users of `allow_in_graph` propagate it to downstream consumer, but in our case, dynamo itself is our consumer.

Reply:
I don't think we should `allow_in_graph` this, I think we should have an annotation like `run_in_dynamo` or a better name which signals to dynamo that it's safe to run this function at trace time.

Reply:
Is this some sort of constant folding thing? Because it looks like constant folding.

Reply:
@mlazos regardless, since `@allow_in_graph`'s contract is actually quite relaxed, I think we can support returning values that are guarded on input tensor metas? (e.g. has_sparse_grad and has_complex). I think what we have now is actually qutie ok.

The part that actually violates `@allow_in_graph`'s contract is actually:
> Your function is functional. That is, it does not mutate any state. This may be relaxed; we actually support functions that appear > to be functional from the outside: they may have in-place PyTorch operations, but may not mutate global state or inputs to the function.

In this case, we need to enforce that there are guards on these mutated values. `_init_group` does obey this. But `@allow_in_dynamo` should somehow tell the compiler to guard the inputs rather than doing it on a case-by-case basis as is done now.

For this purpose, I suggest the following API:

```python
@allow_in_dynamo(mutations=[self.state, self.param_groups])
def _init_group(self, ...):
   ...
```

Then we can guard on the `self.state` and `self.param_groups` using pytree.flatten. Do you agree this is a strong enough contract / good API @ezyang @voznesenskym?

Reply:
If we are ok with this, we can 
1. Orthogonally incrementally refactor `_init_group`'s capture implementation
2. Orthogonally unblock this PR (which is no worse than the status quo), thus unblocking the work related to https://github.com/pytorch/pytorch/pull/110706 that will unlock a lot of performance for Optimizer compilation (it is super slow now - 100s+ on 200 params),

Reply:
Created a proposal for the incremental refactor work: https://github.com/pytorch/pytorch/issues/111331

Reply:
> Is this some sort of constant folding thing? Because it looks like constant folding.

Yeah it's really similar to `assume_constant_result` except with more guarding

Reply:
I'm going to resolve this given that we're now tracking the orthogonal refactor work in https://github.com/pytorch/pytorch/issues/111331